sui-execution/latest/sui-move-natives/src/test_scenario.rs

```diff
@@ -23,21 +24,64 @@ use move_vm_types::{
 };
 use smallvec::smallvec;
 use std::{
-    borrow::Borrow,
+    borrow::{Borrow, BorrowMut},
     collections::{BTreeMap, BTreeSet, VecDeque},
+    sync::Mutex,
 };
 use sui_types::{
     base_types::{ObjectID, SequenceNumber, SuiAddress},
+    digests::{ObjectDigest, TransactionDigest},
+    execution::DynamicallyLoadedObjectMetadata,
     id::UID,
-    object::Owner,
+    in_memory_storage::InMemoryStorage,
+    object::{MoveObject, Object, Owner},
+    storage::ChildObjectResolver,
 };
 
 const E_COULD_NOT_GENERATE_EFFECTS: u64 = 0;
 const E_INVALID_SHARED_OR_IMMUTABLE_USAGE: u64 = 1;
 const E_OBJECT_NOT_FOUND_CODE: u64 = 4;
+const E_UNABLE_TO_ALLOCATE_RECEIVING_TICKET: u64 = 5;
+const E_RECEIVING_TICKET_ALREADY_ALLOCATED: u64 = 6;
+const E_UNABLE_TO_DEALLOCATE_RECEIVING_TICKET: u64 = 7;
 
 type Set<K> = IndexSet<K>;
 
+/// An in-memory test store is a thin wrapper around the in-memory storage in a mutex. The mutex
+/// allows this to be used by both the object runtime (for reading) and the test scenario (for
+/// writing) while hiding mutability.
+#[derive(Tid)]
+pub struct InMemoryTestStore(pub &'static Mutex<InMemoryStorage>);
```

Comment:
I don't quite understand the Mutex here. Can you link to the relevant code? (what goes wrong if it is missing)

Reply:
This is needed because of how this is used. 

In `unit_test.rs` one of these is constructed and placed in `TEST_STORE`. 

The `Mutex` is allowing us to then access the inner `InMemoryStorage` here to place objects to then be received. I'll tag the places where this is used.